### PR TITLE
[server] Broadcast headless updates to subscribers WEB-598, WEB-600

### PR DESCRIPTION
**Pull request by easyCZ**
## Related Issue(s)

* Depends on https://github.com/gitpod-io/gitpod/pull/18214

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - mp-server-bf5c09fbef</li>
	<li><b>🔗 URL</b> - <a href="https://mp-server-bf5c09fbef.preview.gitpod-dev.com/workspaces" target="_blank">mp-server-bf5c09fbef.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - mp-server-headless-updates-gha.13420</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>

/hold

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -115,6 +115,7 @@ jobs:
           mkdir -p "$(dirname "$PREVIEW_ENV_DEV_SA_KEY_PATH")"
           echo "${PREVIEW_ENV_DEV_SA_KEY}" > "${PREVIEW_ENV_DEV_SA_KEY_PATH}"
           gcloud auth activate-service-account --key-file "${PREVIEW_ENV_DEV_SA_KEY_PATH}"
+          gcloud auth configure-docker --quiet
```
Comment (svenefftinge) [inline .github/workflows/build.yml:118]: why is this needed?

File: components/server/src/messaging/redis-subscriber.ts [hunk 2]
```diff
@@ -172,8 +210,13 @@ export class RedisSubscriber implements LocalMessageBroker {
     }
 
     listenForPrebuildUpdatableEvents(listener: HeadlessWorkspaceEventListener): Disposable {
-        // TODO: not implemented
-        return Disposable.create(() => {});
+        // we're being cheap here in re-using a map where it just needs to be a plain array.
```
Comment (svenefftinge) [inline components/server/src/messaging/redis-subscriber.ts:213]: i did not understand that comment as I don't see a map being used.

File: components/server/src/workspace/workspace-starter.ts [hunk 3]
```diff
@@ -796,6 +814,11 @@ export class WorkspaceStarter {
             instance.status.message = `Workspace cannot be started: ${err}`;
             await this.workspaceDb.trace({ span }).storeInstance(instance);
             await this.messagebus.notifyOnInstanceUpdate(workspace.ownerId, instance);
+            await this.publisher.publishInstanceUpdate({
+                instanceID: instance.id,
```
Comment (svenefftinge) [inline components/server/src/workspace/workspace-starter.ts:818]: surprised you still pass the instanceIDher, as I understood it was removed ☝️

Comment (easyCZ) [reply components/server/src/workspace/workspace-starter.ts:818]: For Instance Updates, we need to supply the InstanceID, but for Prebuild Updates, we don't pass the instance ID

Comment (svenefftinge) [reply components/server/src/workspace/workspace-starter.ts:818]: Ah, please ignore the diff was folded.

**Review by svenefftinge — APPROVED** [review]

File: components/server/src/messaging/redis-subscriber.ts (hunk 2 above)
Comment (easyCZ) [reply components/server/src/messaging/redis-subscriber.ts:213]: This is a direct copy-paste from the existing local-messagebus.ts. Basically, we don't track the listeners by an ID for the Prebuild Udpatable, but we re-use a map, with a single key to re-use the helper for creating the listeners. I'll clean this up after we get rid of messagebus, but wanted to keep it as close to the current implementation to reduce chance for mistakes

File: .github/workflows/build.yml (hunk 1 above)
Comment (easyCZ) [reply .github/workflows/build.yml:118]: This is in the parent PR, and just shows here before the rebase. Will be removed once I rebase again
